aocdbc/day_4_passport_processing.py

```diff
@@ -0,0 +1,107 @@
+from re import fullmatch
+import sys
+from typing import Dict, List, Optional
+
+from icontract import require, ensure
+
+_REQUIRED_KEYS = set(["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"])
+
+
+@ensure(lambda: True)
+def blank_line_split(text: str) -> List[str]:
+    # NOTE: I tried to do a lookahead for the 2nd newline, but split doesn't take
+    # a regex.
+    return text.split("\n\n")
+
+
+@ensure(lambda: True)
+def parse(passport: str) -> Optional[Dict[str, str]]:
+    result = {}
+    for tok in passport.split():
+        pair = tok.split(":", 2)
+        if len(pair) != 2:
+            return None
+        result[pair[0]] = pair[1]
+    return result
+
+
+@ensure(lambda entry, result: result == (not (_REQUIRED_KEYS - entry.keys())))
+def is_valid(entry: Dict[str, str]) -> bool:
+    return len(_REQUIRED_KEYS - entry.keys()) == 0
+
+
+@ensure(lambda result: result >= 0)
+def count_valid(batch: str) -> int:
+    count = 0
+    for passport_text in blank_line_split(batch):
+        passport = parse(passport_text)
+        if passport and is_valid(passport):
+            count += 1
+    return count
+
+
+# part 2
+
+
+@ensure(lambda: True)
+def is_valid2(entry: Dict[str, str]) -> bool:
+    if not is_valid(entry):
+        return False
+    for key, value in entry.items():
+        if not is_kv_valid2(key, value):
+            return False
+    return True
+
+
+@ensure(lambda: True)
+def is_kv_valid2(key: str, value: str) -> object:
```

Comment:
`Any` is a bit shorter / easier to read.

```suggestion
def is_kv_valid2(key: str, value: str) -> Any:
```

Reply:
Ah, `Any` and `object` have different meanings, and this choice is intentional. I want `is_kv_valid2(...).something` to create an error, but `if is_kv_valid2(...):` to be ok.

Reply:
A second thought: just to make it easier for the beginners reading the code, maybe you can change the return value to `bool` and add `fullmatch(...) is not None` as a return value?